File: ServerApp/ServerApp/GameState/Player.cs
using System;
using Newtonsoft.Json;

namespace ServerApp.GameState
{
    public class Player: IPlayer
    {
        public string Name { get; }
        public IPosition Position { get; set; }
        public IStats Stats { get; set; }
        public Direction NextMove { get; set; }

        public bool Alive => this.Stats.CurrentHealth > 0;

        public int TakeDamage(int damage)
        {
            Stats.ReduceHealt(damage);
            return damage;
        }
        public int Gold { get; set; }
        public int Experience { get; set; } = 0;

        [JsonConstructor]
        public Player(IPosition position, IStats stats, string name = "Player McName", int gold = 0) 
        {
            Position = position;
            Stats = stats;
            Name = name;
            Gold = gold;
        }

        public void AddExperience(int xp)
        {
            if (xp > 0)
            {
                Experience += xp;
            }
        }

        public void AddGold(int gold)
        {
            if (gold > 0)
            {
                Gold += gold;
            }
        }

        public bool RemoveGold(int gold)
        {
            if (gold < 0) return false;

            if(Gold < gold)
            {
                return false;
            }
            
            Gold -= gold;
            return true;
        }

        public void SetNextMove(Direction direction)
        {
            NextMove = direction;
        }

        public int DropLoot()
        {
            throw new NotImplementedException();
        }
    }
}

File: clientApp/src/GameState/Player.ts
import { IPlayer } from './IPlayer';
import { IPosition } from './IPosition';
import { Position } from './Position';

export class Player implements IPlayer {
    public name: string;
    public position: IPosition;
    public health: number;
    public maxHealth: number;
    public gold: number;
    public experience: number;
    public damage: number;

    constructor(name: string,
                xPos: number,
                yPos: number,
                health: number,
                maxHealth: number,
                gold: number,
                experience: number,
                damage: number) {
        this.name = name;
        this.position = new Position(xPos, yPos);
        this.health = health;
        this.maxHealth = maxHealth;
        this.gold = gold;
        this.experience = experience;
        this.damage = damage;
    }
}
